bluetooth: services: Added CTS service

cx-anuj-pathak opened the pull request:
### Description
This PR attempts to add CTS service to Bluetooth subsystem.

### Changes
- Added CTS implementation and Required KConfig/CMake changes.
- Updated `samples/bluetooth/peripheral` to use this implementation instead of custom one.

Fixes: #77595

cx-anuj-pathak committed:
sys: timeutil: expose and use macro for base year

timeutil internally uses 1900 as base year for calculation,
that is hidden from caller context. Thus declared a macro in
the public header file, and uses that macro will let caller adjust the
calulations accordingly.

Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak committed:
bluetooth: services: added current time service

- added ctc implementation
- updated Kconfig & CMake files
- updated service doc

Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak committed:
samples: bluetooth: peripheral: updated project to use cts from tree

- removed custom cts implementation
- enabled CTS through proj.conf and used it from the tree

Signed-off-by: Anuj Pathak <anuj@croxel.com>

jori-nordic commented on subsys/bluetooth/services/cts.c:
```diff
@@ -0,0 +1,177 @@
+	if (!err) {
+		return len;
+	} else {
+		return BT_GATT_ERR(0x80);
```
what's the 0x80?

cx-anuj-pathak replied:
as per CTS_SPEC_V1.1.0.pdf, section 1.6, server need to send 0x80 error code in case one or more fields are ignored

jori-nordic reviewed (changes requested):
Thanks for the PR ⌚📡

I would prefer if the implementation was not tied to a single backend (here POSIX).

E.g. you could take `struct bt_cts_time_format` as input param to `bt_cts_settime` API.

Similarly, you could allow registering callbacks for the `write` operation, and pass the written `struct bt_cts_time_format` as a param to the callback.

jori-nordic commented on subsys/bluetooth/services/cts.c:
```diff
@@ -0,0 +1,191 @@
+	date_time.tm_sec = ct_time->sec;       /* seconds of minute */
+	date_time.tm_wday = ct_time->wday % 7; /* for sundays convert to 0, else keep same */
+
+	LOG_INF("CTS Write Time: %d/%d/%d %d:%d:%d", date_time.tm_year, date_time.tm_mon,
```
usually those would be `DBG`-level logs

jori-nordic reviewed (commented):
Thanks, the API looks better now. I'll approve as soon as you've split out the HRM changes.

jhedberg commented on subsys/bluetooth/services/Kconfig.cts:
```diff
@@ -0,0 +1,14 @@
+
+if BT_CTS
+
+config BT_CTS_HELPER_API_ENABLE
+	bool "Helper APIs to encode and decode CTS formatted time"
```
I'd drop the `_ENABLE` since it's implicit, and (AFAIK) we don't have a convention of naming boolean Kconfig options this way.

jhedberg commented on include/zephyr/bluetooth/services/cts.h:
```diff
@@ -0,0 +1,143 @@
+ */
+int bt_cts_send_notification(enum bt_cts_update_reason reason);
+
+#if CONFIG_BT_CTS_HELPER_API_ENABLE
```
This is still referencing the old name. That said, I think you can just remove it. Putting API declarations behind #ifdef means that you can't use `if (IS_ENABLED())` when making calls to them.

cx-anuj-pathak replied:
Okay I'll remove them, if `if (IS_ENABLED())` is more preferred way than #ifdef

jhedberg commented on samples/bluetooth/peripheral/prj.conf:
```diff
@@ -11,6 +11,8 @@ CONFIG_BT_ATT_PREPARE_COUNT=5
 CONFIG_BT_HRS=y
 CONFIG_BT_IAS=y
+CONFIG_BT_CTS=y
+CONFIG_BT_CTS_HELPER_API_ENABLE=y
```
Reference to old name here as well - seems like you didn't test the latest change? :)

cx-anuj-pathak replied:
My bad, I have retest latest push.
Thanks!

jhedberg commented on subsys/bluetooth/Kconfig.logging:
```diff
@@ -492,6 +492,12 @@ module-str = BAS
 source "subsys/logging/Kconfig.template.log_config"
 endif # BT_BAS
 
+# CTS
+
+module = BT_CTS
```
This is missing `if BT_CTS`

jhedberg commented on include/zephyr/bluetooth/services/cts.h:
```diff
@@ -0,0 +1,139 @@
+ */
+int bt_cts_send_notification(enum bt_cts_update_reason reason);
+
+/**
+ * @brief Helper API to decode CTS formatted time into milliseconds since epoch
+ *
+ * @param ct_time [IN]  cts time formatted time
+ * @param unix_ms [OUT] pointer to store parsed millisecond since epoch
+ * @return int    0 on success
+ *                negative error codes on failure
+ */
+int bt_cts_time_to_unix_ms(const struct bt_cts_time_format *ct_time, int64_t *unix_ms);
+
+/**
+ * @brief Helper API to encode milliseconds since epoch to CTS formatted time
+ *
+ * @param ct_time [OUT] Pointer to store CTS formatted time
+ * @param unix_ms [IN]  milliseconds since epoch to be converted
+ * @return int    0 on success
+ *                negative error codes on failure
+ */
+int bt_cts_time_from_unix_ms(struct bt_cts_time_format *ct_time, int64_t unix_ms);
```
Might still make sense to mention which Kconfig option these depend on. There's going to be a convenience helper available for it soon, but it hasn't been merged yet: #76783

cx-anuj-pathak replied:
followed settings load reference, but now realised that check is for complete subsystem.
updated comment using reference from `sys_poweroff` API

jhedberg commented on subsys/bluetooth/services/cts.c:
```diff
@@ -0,0 +1,192 @@
+
+static const struct bt_cts_cb *_cb;
+
+#if CONFIG_BT_CTS_HELPER_API
```
This should be #ifdef and not #if. If the Kconfig option is not enabled the macro will not be defined at all.

jhedberg commented on subsys/bluetooth/services/cts.c:
```diff
@@ -0,0 +1,192 @@
+	return 0;
+}
+
+#endif
```
Especially for longer #ifdef..#endif blocks its good practice to put a comment reminding which condition terminates here, i.e. something like:
```c
#endif /* CONFIG_BT_CTS_HELPER_API */
```

jhedberg commented on subsys/bluetooth/services/cts.c:
```diff
@@ -0,0 +1,192 @@
+#define BT_CTS_ATT_ERR_VALUES_IGNORED 0x80
+#define BT_CTS_FRACTION_256_MAX_VALUE 255
+
+static const struct bt_cts_cb *_cb;
```
I'd just call this `cts_cb`. Underscore prefix can be useful e.g. for public struct members where you want to emphasize that it's private, but a static variable like this is implicitly private to the module it's in, so putting the prefix there seems unnecessary.

jhedberg commented on subsys/bluetooth/services/cts.c:
```diff
@@ -0,0 +1,192 @@
+	int err;
+	struct bt_cts_time_format ct_time;
+
+	if (_cb->fill_current_cts_time == NULL) {
+		return -EFAULT;
+	}
```
If it's invalid to register a callback struct with this member set to NULL, why don't you check for it in `bt_cts_init()` and return `-EINVAL` from there? Here in `bt_cts_send_notification()` it seems like it should at most be an `__ASSERT` (if anything at all)

jori-nordic commented:
@Thalley 
```
tests/bsim/bluetooth/audio/test_scripts/csip_forced_release.sh FAILED (30.161 s)
```

This seems unrelated, could you take a look?

jori-nordic commented:
@cx-anuj-pathak The peripheral sample doesn't build

<details>
  <summary>Build logs (expand)</summary>

```
[46/184] Building C object zephyr/CMakeFiles/zephyr.dir/subsys/bluetooth/services/cts.c.obj
506
FAILED: zephyr/CMakeFiles/zephyr.dir/subsys/bluetooth/services/cts.c.obj 
507
ccache /opt/toolchains/zephyr-sdk-0.16.8/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -DPICOLIBC_LONG_LONG_PRINTF_SCANF -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR__=1 -I/__w/zephyr/zephyr/kernel/include -I/__w/zephyr/zephyr/arch/x86/include -I/__w/zephyr/zephyr/twister-out/qemu_x86/samples/bluetooth/peripheral/sample.bluetooth.peripheral/zephyr/include/generated/zephyr -I/__w/zephyr/zephyr/include -I/__w/zephyr/zephyr/twister-out/qemu_x86/samples/bluetooth/peripheral/sample.bluetooth.peripheral/zephyr/include/generated -I/__w/zephyr/zephyr/soc/intel/atom -I/__w/zephyr/zephyr/soc/intel/atom/. -I/__w/zephyr/zephyr/subsys/bluetooth -I/__w/zephyr/zephyr/subsys/settings/include -I/__w/zephyr/modules/crypto/tinycrypt/lib/include -isystem /__w/zephyr/zephyr/lib/libc/common/include -m32  -fno-strict-aliasing -Werror -Os -imacros /__w/zephyr/zephyr/twister-out/qemu_x86/samples/bluetooth/peripheral/sample.bluetooth.peripheral/zephyr/include/generated/zephyr/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -m32 -msoft-float -Wa,--divide --sysroot=/opt/toolchains/zephyr-sdk-0.16.8/x86_64-zephyr-elf/x86_64-zephyr-elf -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wdouble-promotion -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -ftls-model=local-exec -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/samples/bluetooth/peripheral=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -mpreferred-stack-boundary=2 -mno-mmx -mno-sse --specs=picolibc.specs -std=c99 -MD -MT zephyr/CMakeFiles/zephyr.dir/subsys/bluetooth/services/cts.c.obj -MF zephyr/CMakeFiles/zephyr.dir/subsys/bluetooth/services/cts.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/subsys/bluetooth/services/cts.c.obj -c /__w/zephyr/zephyr/subsys/bluetooth/services/cts.c
508
/__w/zephyr/zephyr/subsys/bluetooth/services/cts.c:178:5: error: conflicting types for 'bt_cts_send_notification'; have 'int(uint8_t)' {aka 'int(unsigned char)'}
509
  178 | int bt_cts_send_notification(uint8_t reason)
510
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
511
In file included from /__w/zephyr/zephyr/subsys/bluetooth/services/cts.c:17:
512
/__w/zephyr/zephyr/include/zephyr/bluetooth/services/cts.h:111:5: note: previous declaration of 'bt_cts_send_notification' with type 'int(enum bt_cts_update_reason)'
513
  111 | int bt_cts_send_notification(enum bt_cts_update_reason reason);
514
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
515
ninja: build stopped: subcommand failed.
516

517
```

</details>

jori-nordic commented on include/zephyr/bluetooth/services/cts.h:
```diff
@@ -0,0 +1,145 @@
+int bt_cts_init(const struct bt_cts_cb *cb);
+
+/**
+ * @brief Update CTS Server time and notify all clients
```
You're not updating the time anymore since CTS doesn't have a backend.
You can rephrase to only "Notify all connected clients that have enabled the time update notification"

jori-nordic commented on samples/bluetooth/peripheral/src/main.c:
```diff
@@ -378,9 +425,13 @@ int main(void)
-		cts_notify();
-
+		/* Current time update notification example
+		 * For testing purpose we do update it periodically to send notifications
```
"For testing purposes, we send a manual update notification every second." The rest of the sentence is fine.

cx-anuj-pathak commented on include/zephyr/bluetooth/services/cts.h:
```diff
@@ -0,0 +1,145 @@
+};
+
+/**
+ * @brief This API should be called at boot to ensure proper functioning of CTS service
```
changed it to application init, and it is safe to call it before or after bt_enable.

cx-anuj-pathak commented on include/zephyr/bluetooth/services/cts.h:
```diff
@@ -0,0 +1,145 @@
+/**
+ * @brief Current Time Service (CTS)
+ * @defgroup bt_cts Current Time Service (CTS)
+ * @ingroup bluetooth
```
Added CTS entry. 

observed that dis and nus entry is also missing. and these two didn't have @defgroup defined in header files. May we want to correct and included them too later in separate PR

jori-nordic replied:
yeah that would be great!

cx-anuj-pathak commented:
> @cx-anuj-pathak The peripheral sample doesn't build
> 
> Build logs (expand)

I have just tested it and it builds fine for me.  following are the build logs with git log. I didn't even getting any warning.
May be I'm missing some setup or build step, Will you review it and let me know for future cases

But i can confirm signature mismatch and will correct it with review fixes

```
PS D:\ZephyrApps\zephyr> west build -b nrf52dk/nrf52832 .\samples\bluetooth\peripheral\ -p
-- west build: making build dir D:\ZephyrApps\zephyr\build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: D:/ZephyrApps/zephyr/samples/bluetooth/peripheral
-- CMake version: 3.23.2
-- Found Python3: C:/Python310/python.exe (found suitable version "3.10.5", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: D:/ZephyrApps/zephyr/.cache
-- Zephyr version: 3.7.99 (D:/ZephyrApps/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: nrf52dk, qualifiers: nrf52832
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.8 (D:/ZephyrApps/zephyr-sdk-0.16.8)
-- Found toolchain: zephyr 0.16.8 (D:/ZephyrApps/zephyr-sdk-0.16.8)
-- Found Dtc: C:/ProgramData/chocolatey/bin/dtc.exe (found suitable version "1.5.0", minimum required is "1.4.6") 
-- Found BOARD.dts: D:/ZephyrApps/zephyr/boards/nordic/nrf52dk/nrf52dk_nrf52832.dts
-- Generated zephyr.dts: D:/ZephyrApps/zephyr/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: D:/ZephyrApps/zephyr/build/zephyr/include/generated/zephyr/devicetree_generated.h
-- Including generated dts.cmake file: D:/ZephyrApps/zephyr/build/zephyr/dts.cmake
Parsing D:/ZephyrApps/zephyr/Kconfig
Loaded configuration 'D:/ZephyrApps/zephyr/boards/nordic/nrf52dk/nrf52dk_nrf52832_defconfig'
Merged configuration 'D:/ZephyrApps/zephyr/samples/bluetooth/peripheral/prj.conf'
Configuration saved to 'D:/ZephyrApps/zephyr/build/zephyr/.config'
Kconfig header saved to 'D:/ZephyrApps/zephyr/build/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: d:/zephyrapps/zephyr-sdk-0.16.8/arm-zephyr-eabi/arm-zephyr-eabi/bin/ld.bfd.exe (found version "2.38") 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: D:/ZephyrApps/zephyr-sdk-0.16.8/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc.exe
-- Configuring done
-- Generating done
-- Build files have been written to: D:/ZephyrApps/zephyr/build
-- west build: building application
[1/255] Generating include/generated/zephyr/version.h
-- Zephyr version: 3.7.99 (D:/ZephyrApps/zephyr), build: v3.7.0-1501-g0184674b7940
[255/255] Linking C executable zephyr\zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      186996 B       512 KB     35.67%
             RAM:       30109 B        64 KB     45.94%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from D:/ZephyrApps/zephyr/build/zephyr/zephyr.elf for board: nrf52dk
PS D:\ZephyrApps\zephyr> git log
commit 0184674b7940e0944e122892413471f3ff9e64a6 (HEAD -> subsys/bluetooth/services/cts, origin/subsys/bluetooth/services/cts)
Author: Anuj Pathak <anuj@croxel.com>
Date:   Thu Aug 22 15:31:30 2024 +0530

    samples: bluetooth: peripheral: updated project to use cts from tree
    
    - removed custom cts implementation
    - enabled CTS through proj.conf and used it from the tree
    
    Signed-off-by: Anuj Pathak <anuj@croxel.com>

commit d62b74c07e34793ac5efd287e39506ff38c3bb9b
Author: Anuj Pathak <anuj@croxel.com>
Date:   Thu Aug 22 15:29:41 2024 +0530

    bluetooth: services: added current time service
    
    added ctc implementation and updated Kconfig & CMake files
    
    Signed-off-by: Anuj Pathak <anuj@croxel.com>

commit 483f8bf52ba8f7cdcb59d144194ae6c22f4a9741
Author: Anuj Pathak <anuj@croxel.com>
Date:   Thu Aug 22 15:25:04 2024 +0530

    sys: timeutil: expose and use macro for base year
    
    timeutil internally uses 1900 as base year for calculation,
    that is hidden from caller context. Thus declared a macro in
    the public header file, and uses that macro will let caller adjust the
    calulations accordingly.
    
    Signed-off-by: Anuj Pathak <anuj@croxel.com>
 ```

jori-nordic commented:
I see you have issues with trailing whitespace and whatnot. I recommend the editorconfig extension for your editor. Zephyr will auto-configure your editor to remove them, and also set indentation properly, etc.

If you use vscode, you can just apply these settings (or even use the devcontainer locally if you want):
https://github.com/zephyrproject-rtos/zephyr/pull/77465/files#diff-8298b9fdb00c943c6bc238a1797978ec760110d2b9c51841a2f4615640c94812R19

Thalley commented:
> @Thalley
> 
> ```
> tests/bsim/bluetooth/audio/test_scripts/csip_forced_release.sh FAILED (30.161 s)
> ```
> 
> This seems unrelated, could you take a look?

Just looks like a disconnection to the test server:

> bsim-test
> The self-hosted runner: zrv2-linux-x64-4xlarge-cnx-bwqtb-runner-j9q88 lost communication with the server. Verify the machine is running and has a healthy network connection. Anything in your workflow that terminates the runner process, starves it for CPU/Memory, or blocks its network access can cause this error.

Retriggered

jori-nordic commented:
@cx-anuj-pathak 

<details>
  <summary>Doc build is failing</summary>

```
[0/2] Generating Devicetree bindings documentation...
[1/2] Running Sphinx HTML build...
/home/runner/_work/zephyr/zephyr/doc/_build/src/connectivity/bluetooth/api/services.rst:12: WARNING: Title underline too short.

Current Time Service
***************
/home/runner/_work/zephyr/zephyr/doc/_build/src/connectivity/bluetooth/api/services.rst:12: WARNING: Title underline too short.

Current Time Service
***************
FAILED: CMakeFiles/html 
cd /home/runner/_work/zephyr/zephyr/doc/_build && /home/runner/.local/lib/python3.10/site-packages/cmake/data/bin/cmake -E env DOXYGEN_EXECUTABLE=/opt/doxygen-1.12.0/bin/doxygen DOT_EXECUTABLE=/usr/bin/dot DOCS_HTML_DIR=/home/runner/_work/zephyr/zephyr/doc/_build/html /home/runner/.local/bin/sphinx-build -b html -c /home/runner/_work/zephyr/zephyr/doc -d /home/runner/_work/zephyr/zephyr/doc/_build/doctrees -w /home/runner/_work/zephyr/zephyr/doc/_build/html.log -t development -j 14 -W --keep-going -T -q -t publish /home/runner/_work/zephyr/zephyr/doc/_build/src /home/runner/_work/zephyr/zephyr/doc/_build/html
```

</details>
